include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -35,25 +35,79 @@ class sycl_backend
     using execution_resource_t = resource_type;
     using resource_container_t = std::vector<execution_resource_t>;
 
+    bool has_enable_profiling = false;
   private:
-    class async_waiter
+    class async_waiter_base{
+        public:
+            virtual void wait() = 0;
+            virtual void report() = 0;
+            virtual bool is_complete() = 0;
+    };
+
+    template<typename Selection>
+    class async_waiter : public async_waiter_base
     {
         sycl::event e_;
-
+        Selection* s;
       public:
-        async_waiter(const sycl::event& e) : e_(e) {}
+        async_waiter(sycl::event e) : e_(e){}
+        async_waiter(sycl::event e, Selection* selection) : e_(e), s(selection) {}
+
         sycl::event
         unwrap()
         {
             return e_;
         }
+
         void
-        wait()
+        wait() override
         {
             e_.wait();
         }
+
+        void
+        report() override{
+            if constexpr (report_value_v<Selection, execution_info::task_time_t>){
+                cl_ulong time_start = e_.template get_profiling_info<sycl::info::event_profiling::command_start>();
+                cl_ulong time_end = e_.template get_profiling_info<sycl::info::event_profiling::command_end>();
+                s->report(execution_info::task_time, time_end-time_start);
+            }
+
+        }
+
+        bool
+        is_complete() override{
+            return e_.get_info<sycl::info::event::command_execution_status>() == sycl::info::event_command_status::complete;
+        }
+
     };
 
+    struct async_waiter_list_t{
+
+        std::mutex m_;
+        std::vector<async_waiter_base*> async_waiters;
+
+        template<typename T>
+        void add_waiter(T *t){
+            std::lock_guard<std::mutex> l(m_);
+            async_waiters.push_back(t);
+        }
+
+        void lazy_report(){
+            std::lock_guard<std::mutex> l(m_);
+            int size = async_waiters.size();
+            for(auto i = async_waiters.begin(); i!=async_waiters.begin()+size; i++){
+                if((*i)->is_complete()){
+                    (*i)->report();
+                    async_waiters.erase(i);
```

Comment:
...because you simple delete item from vector without `delete` call.

Reply:
Fixed. Used a std::unique_ptr